[DOC-1300][2026.1] YugabyteDB Flink CDC support (#31798)
yugabyte/yugabyte-db: docs/content/stable/additional-features/change-data-capture/using-flink-cdc/get-started.md

    parent: explore-change-data-capture-flink-cdc
    weight: 10
type: docs
---

This tutorial walks through deploying a YugabyteDB-to-PostgreSQL change data capture pipeline using Docker Compose and the Flink SQL Client. You should have a running Flink job that captures INSERT, UPDATE, and DELETE events from a YugabyteDB table and applies them to a PostgreSQL sink table in real time.

## Prerequisites

Before you begin, ensure you have the following:

- Apache Flink 1.20.x cluster (validated version) with the `postgres-cdc` and JDBC sink connector JARs placed in `/opt/flink/lib/`.

    Pre-built JARs for YugabyteDB are published on the [Yugabyte Flink CDC Releases page](https://github.com/yugabyte/flink-cdc/releases). The Docker image used in this tutorial bundles them automatically.

- YugabyteDB cluster with [YSQL logical replication enabled](../../../../explore/change-data-capture/#set-up-pg-recvlogical) and network connectivity to your Flink cluster.

    Note the IP address of a YB-TServer node that is reachable from the Flink containers.

- [Docker](https://docs.docker.com/) and [Docker Compose](https://docs.docker.com/compose/) installed and running.

> - A sink PostgreSQL database accessible from the Flink containers.

Reviewer: ```suggestion
- A PostgreSQL sink database accessible from the Flink containers.
```

## Configure the source

Connect to your YugabyteDB cluster using `ysqlsh` and run the following SQL to create the source table, publication, and replication slot:

```sql
CREATE TABLE shipments (
  shipment_id INT PRIMARY KEY,
  order_id    INT,
  origin      TEXT,
  destination TEXT,
  is_arrived  BOOLEAN
);

INSERT INTO shipments VALUES
  (1001, 1, 'Beijing',  'Shanghai',    TRUE),
  (1002, 2, 'New York', 'Los Angeles', FALSE),
  (1003, 3, 'Mumbai',   'Delhi',       TRUE);

CREATE PUBLICATION dbz_publication FOR ALL TABLES;
SELECT * FROM pg_create_logical_replication_slot('flink', 'pgoutput');
```

> {{< note title="Slot name uniqueness" >}}

Reviewer: ```suggestion
{{< note title="Use unique slot names" >}}
```

> Assign a unique `slot.name` to each Flink pipeline. Using the same slot name in multiple pipelines causes errors about active PIDs on the same slot.

Reviewer: ```suggestion
Assign a unique `slot.name` to each Flink pipeline. Using duplicate slot names in multiple pipelines will result in errors about active PIDs on the same slot.
```

{{< /note >}}

## Initialize the target database

Connect to your PostgreSQL sink database and create the target table:

```sql
CREATE TABLE shipments (
  shipment_id INT PRIMARY KEY,
  order_id    INT,
  origin      TEXT,
  destination TEXT,
  is_arrived  BOOLEAN
);
```

## Configure Docker Compose

Create a `docker-compose.yaml` file with the following content. Replace the environment variable placeholders with your actual values.

```yaml
services:
  jobmanager:

>     image: <Docker Image>

Author: This seems valid too.

    extra_hosts:
      - "yb-ysql:${YB_YSQL_HOST}"

  taskmanager:
    image: <Docker Image>
    container_name: flink-taskmanager
    hostname: taskmanager
    depends_on: [jobmanager]
    command: taskmanager
    volumes:
      - ./checkpoints:/opt/flink/checkpoints
    environment:
      JOB_MANAGER_RPC_ADDRESS: jobmanager
      TASK_MANAGER_NUMBER_OF_TASK_SLOTS: "4"
      YB_YSQL_HOST: ${YB_YSQL_HOST}
      YB_YSQL_PORT: ${YB_YSQL_PORT}
      SINK_JDBC_URL: ${SINK_JDBC_URL}
      FLINK_PROPERTIES: |-
        restart-strategy.type: fixed-delay
        restart-strategy.fixed-delay.attempts: 100
        restart-strategy.fixed-delay.delay: 15 s
        state.checkpoints.dir: file:///opt/flink/checkpoints
    extra_hosts:
      - "yb-ysql:${YB_YSQL_HOST}"
```

Create a `.env` file in the same directory with the following configuration variables:

```sh
YB_YSQL_HOST=<tserver-ip>
YB_YSQL_PORT=5433
SINK_JDBC_URL=jdbc:postgresql://host.docker.internal:5432/postgres?user=postgres&password=postgres
```

Start the Flink cluster:

```sh
docker compose up -d
```

Verify that both containers are running and the Flink Web UI is accessible at `http://localhost:8081`.

> ## Initiate the streaming job

Reviewer: ```suggestion
## Start the streaming job
```

Start the Flink SQL Client inside the jobmanager container:

```sh
docker compose exec jobmanager ./bin/sql-client.sh
```

> In the SQL Client, configure the runtime and checkpointing settings, then define the source and sink tables:

Reviewer: SQL Client - do you mean ysqlsh as started previously?

```sql
SET 'execution.runtime-mode'                      = 'streaming';
SET 'execution.checkpointing.interval'            = '60 s';
SET 'execution.checkpointing.timeout'             = '10 min';

> -- Source table: YugabyteDB via postgres-cdc connector

Reviewer: usually we split these into steps rather than one big statement block with inline comments

CREATE TABLE yb_shipments (
  shipment_id INT,
  order_id    INT,
  origin      STRING,
  destination STRING,
  is_arrived  BOOLEAN,
  PRIMARY KEY (shipment_id) NOT ENFORCED
) WITH (
  'connector'              = 'postgres-cdc',

>   'hostname'               = '<tserver-ip>',

Author: I think this is a valid change @bakul-gupta - Yes?

>   'hostname'               = '<tserver-ip>',

Reviewer 2: No. In the docker environment you will have to use `host.docker.internal` and for local environment `localhost` or 1`27.0.0.1`. We can mention that replace the `<tserver-ip>` with your host ip and give the above examples.
